Pcap++/header/KniDevice.h

```diff
@@ -0,0 +1,709 @@
+	 *  - On old versions of DPDK only 4 KNI devices may be created:
+	 *    Yep this is a limitation. You can change MAX_KNI_INTERFACES constant in KniDevice.cpp to unlock this limit;
+	 *  - Any set* method never succeeds:
+	 *    You may forgot that they generate KNI requests that Your application MUST handle.
+	 *    Just use KniDevice#startRequestHandlerThread to handle all requests automatically.
+	 *    Or user running the application don't have suitable access rights (must have CAP_NET_ADMIN).
+	 * Usefull links:
+	 *  - <a href="https://doc.dpdk.org/guides/prog_guide/kernel_nic_interface.html">KNI interface concept DPDK documentation</a>
+	 *  - <a href="https://doc.dpdk.org/guides/nics/kni.html">KNI PMD</a>
+	 *  - <a href="https://doc.dpdk.org/guides/sample_app_ug/kernel_nic_interface.html">KNI DPDK sample application</a>
+	 *  - <a href="https://doc.dpdk.org/dts/test_plans/kni_test_plan.html">KNI DPDK test plan</a>
+	 */
+	class KniDevice : public IDevice
+	{
+		friend class MBufRawPacket;
+	public:
+		/**
+		 * Various link related constants for KNI device
+		 */
+		enum KniLinkState
+		{
+			/** Returned by KNI functions if DPDK version used don't support link setup capability */
+			LINK_NOT_SUPPORTED = -2,
+			/** Returned by KNI functions if link changing function meets an error */
+			LINK_ERROR = -1,
+			/** Used to put link status on KNI device DOWN */
+			LINK_DOWN = 0,
+			/** Used to put link status on KNI device UP */
+			LINK_UP = 1
+		};
+		/**
+		 * Various information related constants for KNI device
+		 */
+		enum KniInfoState
+		{
+			/** Used to identify intent to obtain cached version of KNI device information */
+			INFO_CACHED = 0,
+			/** Used to identify intent to renew/update KNI device information */
+			INFO_RENEW = 1
+		};
+		/**
+		 * Promiscuous mode related constants for KNI device
+		 */
+		enum KniPromiscuousMode
+		{
+			/** Used to DISABLE promiscuous mode on KNI device */
+			PROMISC_DISABLE = 0,
+			/** Used to ENABLE promiscuous mode on KNI device */
+			PROMISC_ENABLE = 1
+		};
+		/**
+		 * Callback related constants for KNI device
+		 */
+		enum KniCallbackVersion
+		{
+			/** Reports that DPDK supports only KniIoctlCallbacks callback structure */
+			CALLBACKS_NEW = 0,
+			/** Reports that DPDK supports only KniOldIoctlCallbacks callback structure */
+			CALLBACKS_OLD = 1
+		};
+		/**
+		 * Various callback types supported by KNI device
+		 */
+		enum KniCallbackType
+		{
+			/** KniIoctlCallbacks#change_mtu and KniOldIoctlCallbacks#change_mtu callback */
+			CALLBACK_MTU,
+			/** KniIoctlCallbacks#config_network_if and KniOldIoctlCallbacks#config_network_if callback */
+			CALLBACK_LINK,
+			/** KniIoctlCallbacks#config_mac_address callback */
+			CALLBACK_MAC,
+			/** KniIoctlCallbacks#config_promiscusity callback */
+			CALLBACK_PROMISC
+		};
+
+		/**
+		 * @brief New callbacks for KNI device events.
```

Comment:
This is a KNI device design used by DPDK. User(application) of KNI device must respond to some range of events that are happening asynchronously from system user like changing the MAC, MTU, link state and promisc state of interface (in future releases of DPDK there ma be more). This events are generated by linux kernel thread that handels this KNI device. This thread will be bloacked for at most 3 seconds before it assumes that event is not handled and reports a failure to linux API that have made a call (like ip commad). So this callbacks are called for this device each time when this event is happend. But the API thru which the event will arrive is hiden by DPDK implementation and must be called thru special function where the callbacks are called. I have implemented 2 ways of handling this DPDK API: synchronous (via call to handleRequests) and asynchronous (via startRequestHandlerThread)
In the end this API must be called or the device would not function properly

Reply:
The functions that changes this parameter represents simple calls to Linux API to do same as ip command

Reply:
thanks for making this much more clearer. One thing I still not understand is what do you expect the user to do in these callback in order for the device to operate properly. Is there a default behavior you can implement?

Reply:
As stated in callback documentation (and can be seen in tests) the simplest behavior of a callback is to always return 0 ignoring the callback params to specify for DPDK that event is handled correctly.
Despite that the DPDK will take some actions by itself if the callback is not provided (as I have described here: https://github.com/seladb/PcapPlusPlus/pull/196#discussion_r286219182) and this is a bigger concern.
My approach is to pass to end user the decision on what to do in case of this events and to be transparent with DPDK behaviour (as it may change in future releases).

Reply:
> Is there a default behavior you can implement?

I think not

Reply:
ok, I agree. Please leave it as it is

Reply:
I will review the dcumentation to state more accurately the way that KNI devices work. This conversation will help. Thx.

Reply:
\+